crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -0,0 +1,256 @@
+use bitflags::bitflags;
+
+use ruff_text_size::TextSize;
+
+bitflags! {
+    /// [String and Bytes literals]: https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals
+    /// [PEP 701]: https://peps.python.org/pep-0701/
+    #[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash)]
+    struct StringFlagsInner: u8 {
+        /// The string uses double quotes (`"`).
+        /// If this flag is not set, the string uses single quotes (`'`).
+        const DOUBLE = 1 << 0;
+
+        /// The string is triple-quoted:
+        /// it begins and ends with three consecutive quote characters.
+        const TRIPLE_QUOTED = 1 << 1;
+
+        /// The string has a `u` or `U` prefix.
+        /// While this prefix is a no-op at runtime,
+        /// strings with this prefix can have no other prefixes set.
+        const U_PREFIX = 1 << 2;
+
+        /// The string has a `b` or `B` prefix.
+        /// This means that the string is a sequence of `int`s at runtime,
+        /// rather than a sequence of `str`s.
+        /// Strings with this flag can also be raw strings,
+        /// but can have no other prefixes.
+        const B_PREFIX = 1 << 3;
+
+        /// The string has a `f` or `F` prefix, meaning it is an f-string.
+        /// F-strings can also be raw strings,
+        /// but can have no other prefixes.
+        const F_PREFIX = 1 << 4;
+
+        /// The string has an `r` or `R` prefix, meaning it is a raw string.
+        /// F-strings and byte-strings can be raw,
+        /// as can strings with no other prefixes.
+        /// U-strings cannot be raw.
+        const R_PREFIX = 1 << 5;
+    }
+}
+
+impl TryFrom<&char> for StringFlagsInner {
+    type Error = String;
+
+    fn try_from(value: &char) -> Result<Self, String> {
+        let result = match value {
+            '\'' => Self::empty(),
+            '"' => Self::DOUBLE,
+            'r' | 'R' => Self::R_PREFIX,
+            'u' | 'U' => Self::U_PREFIX,
+            'b' | 'B' => Self::B_PREFIX,
+            'f' | 'F' => Self::F_PREFIX,
+            _ => return Err(format!("Unexpected prefix or quote '{value}'")),
+        };
+        Ok(result)
+    }
+}
+
+impl TryFrom<char> for StringFlagsInner {
+    type Error = String;
+
+    fn try_from(value: char) -> Result<Self, String> {
+        Self::try_from(&value)
+    }
+}
+
+const DISALLOWED_USTRING_PREFIXES: StringFlagsInner = StringFlagsInner::B_PREFIX
```

Comment:
Nit: I'd expect these to be cased as `DISALLOWED_U_STRING_PREFIXES`, to match `Self::U_PREFIX` and (e.g.) "f-string".